README.md
MDBench
=======

A mini-app to represent the multipole resonance representation lookup and Doppler broadening cross section algorithm.


CHANGES
=====================================================================
NEW IN VERSION 11
=====================================================================
- (Feature) Added in an option to
  use an event based simulation algorithm as opposed to the default
  history based algorithm. This same change was featured in the
  XSBench v18 update. This mode can be enabled with the "-m
  event" command line option. The central difference between the
  default history based algorithm and the event based algorithm is
  the dependence or independence of macroscopic cross section
  lookups. In the default history based method, the simulation
  parallelism is expressed over a large number of independent
  particle "histories", with each particle history requiring (by
  default) 34 sequential and dependent macroscopic cross section
  lookups to be executed. In the event based method, all macroscopic
  cross sections are pooled together in a way that allows them to
  be executed independently in any order. On CPU architectures,
  both algorithms run in approximately the same speed, though in
  full Monte Carlo applications there are some added costs associated
  with the event based method (namely, the need to frequently store
  and load particle states) that are not represented in XSBench.
  However, the model event based algorithm in XSBench is useful for
  examining how the dependence or independence of the loop over
  macroscopic cross sections may affect performance on different
  computer architectures.

  The event based mode in RSBench is very similar to the algorithm
  expressed in v9 and before, so performance results collected
  with v9 and previous should be comparable to performance results
  collected with the event mode in v11. On most CPU architectures
  and with most compilers, the history and event based methods run
  in approximately the same time, with some CPUs showing a small
  (less than 20%) speedup with the event based method.

- With the addition of the different event and history based options,
  the main function of the code has been broken down into some
  smaller functions. Notably, the main parallel simulation phase
  has been moved to a history and event based function in the
  "simulation.c" file.

- Added different verification checksums for the new event based mode.
  While in a real application, the history and event based methods
  would produce the same answers, in the context of this mini-app
  for programming convenience purposes they do not produce the same
  answer/checksum.

=====================================================================
NEW IN VERSION 10
=====================================================================
- In previous versions of RSBench, it was assumed that the method
  of parallelism in the app would not be altered, as the app was
  written to respect some loop dependencies of fully featured MC
  applications. However, to foster more experimentation with
  different methods of parallelizing the XS lookup kernel, we have
  decided to add an extra loop (over particles) into RSBench so that
  loop dependencies can be explicit in RSBench. This should make it
  easier for people to optimize or port RSBench without worry that
  they are breaking any implicit loop dependencies. All loop 
  dependencies should now be apparent without requiring any knowledge
  of full MC applications.

  From a performance standpoint, the change does not really affect
  anything regarding default performance, so performance results for
  CPUs (run without modification of the source code) should be
  virtually identical between v9 and v10.

  RSBench now has its default OpenMP thread level parallelism
  expressed over independent particle histories. Each particle
  then executes 34 macro_xs lookups, which are dependent, meaning
  that this loop cannot be parallelized as each iteration is
  dependent on results from the previous iteration. For each
  macro_xs, there is an independent inner loop over micro_xs lookups,
  which is not parallelized by default in RSBench, but could be
  if desired provided that atomic or controlled writes are used
  for the reduction of the macroscopic XS vector.

  The introduction of particles into RSBench follows a similar
  change made in v17 of XSBench.

- (Performance) To re-iterate, there is no expected performance
  change for CPU architectures running the default code. The addition
  of the particle loop into the code was done only to allow those
  altering the code to more transparently see loop dependencies as
  to avoid parallelizing or pre-processing loops in a manner which
  would not be possible in a real MC app.

  There is a slight change in performance in the smaller problem
  size due to the OpenMP dynamic work chunk size. In v9, this was
  set as 1 macro_xs lookup per chunk. In v10, as we are now
  parallelizing over particle histories, the minimum chunk size
  is now effectively 34 macro_xs lookups. For the large problem,
  the thread work scheduling overhead is small compared to the
  cost of a macro_xs lookup, so there isn't any difference. It's
  just in the small problem size where the scheduling overhead
  begins to be significant so the performance difference between
  v9 and v10 shows up. v9 can be made equivalent in performance
  to v10 by simply increasing the dynamic chunk size on the
  macro_xs OpenMP loop.

- (Option) In support of this change, the user now has the ability
  to alter the number of particle histories they want to simulate.
  This can be controlled with the "-p" command line argument.
  By default, 300,000 particle histories will be simulated.
  This is now the recommended argument for users to adjust if
  the want to increase/decrease the runtime of the problem. Real
  MC simulations may use anywhere from O(10^5) to O(10^10)
  particles per power iteration, so a wide range of values here
  is acceptable. 

- (Option) The number of lookups to perform per particle history.

src/io.c
#include "rsbench.h"

// Prints program logo
void logo(int version)
{
	border_print();
	printf(
"                    _____   _____ ____                  _     \n"
"                   |  __ \\ / ____|  _ \\                | |    \n"
"                   | |__) | (___ | |_) | ___ _ __   ___| |__  \n"
"                   |  _  / \\___ \\|  _ < / _ \\ '_ \\ / __| '_ \\ \n"
"                   | | \\ \\ ____) | |_) |  __/ | | | (__| | | |\n"
"                   |_|  \\_\\_____/|____/ \\___|_| |_|\\___|_| |_|\n"
	);
	border_print();
	center_print("Developed at Argonne National Laboratory", 79);
	char v[100];
	sprintf(v, "Version: %d", version);
	center_print(v, 79);
	border_print();
}

// Prints Section titles in center of 80 char terminal
void center_print(const char *s, int width)
{
	int length = strlen(s);
	int i;
	for (i=0; i<=(width-length)/2; i++) {
		fputs(" ", stdout);
	}
	fputs(s, stdout);
	fputs("\n", stdout);
}

void border_print(void)
{
	printf(
	"==================================================================="
	"=============\n");
}

// Prints comma separated integers - for ease of reading
void fancy_int( int a )
{
    if( a < 1000 )
        printf("%d\n",a);

    else if( a >= 1000 && a < 1000000 )
        printf("%d,%03d\n", a / 1000, a % 1000);

    else if( a >= 1000000 && a < 1000000000 )
        printf("%d,%03d,%03d\n", a / 1000000, (a % 1000000) / 1000, a % 1000 );

    else if( a >= 1000000000 )
        printf("%d,%03d,%03d,%03d\n",
               a / 1000000000,
               (a % 1000000000) / 1000000,
               (a % 1000000) / 1000,
               a % 1000 );
    else
        printf("%d\n",a);
}

Input read_CLI( int argc, char * argv[] )
{
	Input input;

	// defaults to the history based simulation method
	input.simulation_method = HISTORY_BASED;
	// defaults to max threads on the system	
	input.nthreads = omp_get_num_procs();
	// defaults to 355 (corresponding to H-M Large benchmark)
	input.n_nuclides = 355;
	// defaults to 300,000
	input.particles = 300000;
	// defaults to 34
	input.lookups = 34;
	// defaults to H-M Large benchmark
	input.HM = LARGE;
	// defaults to 3000 resonancs (avg) per nuclide
	input.avg_n_poles = 1000;
	// defaults to 100
	input.avg_n_windows = 100;
	// defaults to 4;
	input.numL = 4;
	// defaults to no temperature dependence (Doppler broadening)
	input.doppler = 1;
	
	int default_lookups = 1;
	int default_particles = 1;

	// Collect Raw Input
	for( int i = 1; i < argc; i++ )
	{
		char * arg = argv[i];

		// nthreads (-t)
		if( strcmp(arg, "-t") == 0 )
		{
			if( ++i < argc )
				input.nthreads = atoi(argv[i]);
			else
				print_CLI_error();
		}
		// Simulation Method (-m)
		else if( strcmp(arg, "-m") == 0 )
		{
			char * sim_type;
			if( ++i < argc )
				sim_type = argv[i];
			else
				print_CLI_error();

			if( strcmp(sim_type, "history") == 0 )
				input.simulation_method = HISTORY_BASED;
			else if( strcmp(sim_type, "event") == 0 )
			{
				input.simulation_method = EVENT_BASED;
				// Also resets default # of lookups
				if( default_lookups && default_particles )
				{
					input.lookups =  input.lookups * input.particles;
					input.particles = 0;
				}
			}
			else
				print_CLI_error();
		}
		// lookups (-l)
		else if( strcmp(arg, "-l") == 0 )
		{
			if( ++i < argc )
			{
				input.lookups = atoi(argv[i]);
				default_lookups = 0;
			}
			else
				print_CLI_error();
		}
		// particles (-p)
		else if( strcmp(arg, "-p") == 0 )
		{
			if( ++i < argc )
			{
				input.particles = atoi(argv[i]);
				default_particles = 0;
			}
			else
				print_CLI_error();
		}
		// nuclides (-n)
		else if( strcmp(arg, "-n") == 0 )
		{
			if( ++i < argc )
				input.n_nuclides = atoi(argv[i]);
			else
				print_CLI_error();
		}
		// HM (-s)
		else if( strcmp(arg, "-s") == 0 )
		{	
			if( ++i < argc )
			{
				if( strcmp(argv[i], "small") == 0 )
					input.HM = SMALL;
				else if ( strcmp(argv[i], "large") == 0 )
					input.HM = LARGE;
				else
					print_CLI_error();
			}
			else
				print_CLI_error();
		}
		// Doppler Broadening (Temperature Dependence)
		else if( strcmp(arg, "-d") == 0 )
		{	
			input.doppler = 0;
		}
		// Avg number of windows per nuclide (-w)
		else if( strcmp(arg, "-W") == 0 )
		{
			if( ++i < argc )
				input.avg_n_windows = atoi(argv[i]);
			else
				print_CLI_error();
		}
		// Avg number of poles per nuclide (-p)
		else if( strcmp(arg, "-P") == 0 )
		{
			if( ++i < argc )
				input.avg_n_poles = atoi(argv[i]);
			else
				print_CLI_error();
		}
		else
			print_CLI_error();
	}

	// Validate Input

	// Validate nthreads
	if( input.nthreads < 1 )
		print_CLI_error();
	
	// Validate n_isotopes
	if( input.n_nuclides < 1 )
		print_CLI_error();
	
	// Validate lookups
	if( input.lookups < 1 )
		print_CLI_error();
	
	// Validate lookups
	if( input.avg_n_poles < 1 )
		print_CLI_error();
	
	// Validate lookups
	if( input.avg_n_windows < 1 )
		print_CLI_error();
	
	// Set HM size specific parameters
	// (defaults to large)
	if( input.HM == SMALL )
		input.n_nuclides = 68;

	// Return input struct
	return input;
}

void print_CLI_error(void)
{
	printf("Usage: ./multibench <options>\n");
	printf("Options include:\n");
	printf("  -t <threads>            Number of OpenMP threads to run\n");
	printf("  -m <simulation method>  Simulation method (history, event)\n");
	printf("  -s <size>               Size of H-M Benchmark to run (small, large)\n");
	printf("  -l <lookups>            Number of Cross-section (XS) lookups per particle history\n");
	printf("  -p <particles>          Number of particle histories\n");
	printf("  -P <poles>              Average Number of Poles per Nuclide\n");
	printf("  -W <poles>              Average Number of Windows per Nuclide\n");
	printf("  -d                      Disables Temperature Dependence (Doppler Broadening)\n");
	printf("Default is equivalent to: -s large -m history -l 34 -p 300000 -P 1000 -W 100\n");
	printf("See readme for full description of default run values\n");
	exit(4);
}

void print_input_summary(Input input)
{
	// Calculate Estimate of Memory Usage
	size_t mem = get_mem_estimate(input);

	if( input.simulation_method == EVENT_BASED )
		printf("Simulation Method:           Event Based\n");
	else
		printf("Simulation Method:           History Based\n");
	printf("Materials:                   12\n");
	printf("H-M Benchmark Size:          ");
	if( input.HM == 0 )
		printf("Small\n");
	else
		printf("Large\n");
	if( input.doppler == 1 )
		printf("Temperature Dependence:      ON\n");
	else
		printf("Temperature Dependence:      OFF\n");
	printf("Total Nuclides:              %d\n", input.n_nuclides);
	printf("Avg Poles per Nuclide:       "); fancy_int(input.avg_n_poles);
	printf("Avg Windows per Nuclide:     "); fancy_int(input.avg_n_windows);

	int lookups = input.lookups;
	if( input.simulation_method == HISTORY_BASED )
	{
		printf("Particles:                   "); fancy_int(input.particles);
		printf("XS Lookups per Particle:     "); fancy_int(input.lookups);
		lookups *= input.particles;
	}
	printf("Total XS Lookups:            "); fancy_int(lookups);
	printf("Threads:                     %d\n", input.nthreads);
	printf("Est. Memory Usage (MB):      %.1lf\n", mem / 1024.0 / 1024.0);
	#ifdef PAPI
	printf("PAPI Performance Counters:   ON\n");
	#endif
}
